tests/integration_tests/base.py

```diff
@@ -0,0 +1,194 @@
+import functools
+import json
+import re
+from dataclasses import dataclass, field
+from typing import Any, Callable, Dict, Iterator, List
+
+from openai import NOT_GIVEN, NotGiven
+from openai.types.chat import (
+    ChatCompletionMessageParam,
+    ChatCompletionToolParam,
+)
+from openai.types.chat.completion_create_params import Function
+from pydantic import BaseModel
+
+from aidial_adapter_openai.app_config import ApplicationConfig
+from aidial_adapter_openai.constant import ChatCompletionDeploymentType
+from tests.utils.openai import ChatCompletionResult, ExpectedException
+
+
+class DeploymentConfig(BaseModel):
+    deployment_id: str
+    deployment_type: ChatCompletionDeploymentType
+    upstream_endpoint: str
+    upstream_api_key: str
+
+    @property
+    def upstream_headers(self) -> Dict[str, str]:
+        return {
+            "X-UPSTREAM-KEY": self.upstream_api_key,
+            "X-UPSTREAM-ENDPOINT": self.upstream_endpoint,
+        }
+
+
+class TestDeployments(BaseModel):
+    __test__ = False
+    deployments: list[DeploymentConfig]
+
+    @classmethod
+    def from_config(cls, config_path: str):
+        with open(config_path, "r") as f:
+            test_configs = json.load(f)
+
+        return cls(
+            deployments=[DeploymentConfig(**config) for config in test_configs]
+        )
+
+    @property
+    def app_config(self) -> ApplicationConfig:
+        config = ApplicationConfig()
+        for deployment in self.deployments:
+            config.add_deployment(
+                deployment.deployment_id, deployment.deployment_type
+            )
+        return config
+
+
+def sanitize_id_part(value: Any) -> str:
+    """Convert any value to a pytest-safe identifier part."""
+    if isinstance(value, bool):
+        return "on" if value else "off"
+    if isinstance(value, (int, float)):
+        return str(value).replace(".", "p")  # e.g., 0.5 -> 0p5
+    if value is None:
+        return "none"
+
+    value_str = str(value).lower()
+    sanitized = "".join(c if c.isalnum() else "_" for c in value_str)
+    return re.sub("_+", "_", sanitized.strip("_"))
+
+
+@dataclass
+class TestCase:
+    __test__ = False
+
+    deployment_config: DeploymentConfig
+
+    name: str
+    streaming: bool
+
+    messages: List[ChatCompletionMessageParam]
+
+    expected: Callable[[ChatCompletionResult], bool] | ExpectedException
+
+    max_tokens: int | NotGiven
+    stop: List[str] | NotGiven
+
+    n: int | NotGiven
+
+    functions: List[Function] | NotGiven
+    tools: List[ChatCompletionToolParam] | NotGiven
+    temperature: float | NotGiven
+
+    def get_id(self):
+        parts = [
+            sanitize_id_part(self.name),
+            f"{sanitize_id_part(self.deployment_config.deployment_id)}",
```

Comment:
How to distinguish test_case with the same deployment id but **different upstreams**?
Need to print an index of the deployment in the original config to disambiguate the deployments.

Reply:
Yeah, I've included upstream index into test name, now it looks like that:
```
test_chat_completion[image_in_custom_content/GPT4_VISION__gpt_4_vision_preview_upstream_1/stream:off/tokens:800]
```